Auto system theme detection

temidaradev opened the pull request:
This could be a solution to this -> #79 
I added checker for all three specific operating systems, checkers work properly because of special builds for each OS
```go
//go:build darwin
// +build darwin
```
```go
//go:build windows
// +build windows
```
```go
//go:build linux
// +build linux
```
So there is no conflict, so it works properly for each system. But there is no live updated themes like while using the app and you wanted to change the theme but when you change it wont apply to the app.

realskyquest commented:
the .idea files are commited, im sure they are not supposed to be there

temidaradev commented:
oh nooo im so sorry

hajimehoshi commented on context.go:
```diff
@@ -17,17 +18,24 @@ import (
 func init() {
 	// TODO: Consider the system color mode.
 	switch mode := os.Getenv("GUIGUI_COLOR_MODE"); mode {
-	case "light", "":
```
Why did you remove this?

hajimehoshi replied:
Ah, yes, when this is emtpy, this should be fallback to the system theme.

hajimehoshi commented on context.go:
```diff
@@ -5,6 +5,7 @@ package guigui
 
 import (
 	"fmt"
+	"github.com/hajimehoshi/guigui/internal/theme"
```
Reorder the import. See the other files.

hajimehoshi commented on context.go:
```diff
@@ -99,10 +107,9 @@ func (c *Context) ColorMode() ColorMode {
 }
 
 func (c *Context) SetColorMode(mode ColorMode) {
-	if c.hasColorMode && mode == c.colorMode {
```
?

hajimehoshi replied:
ping: what's this change?

temidaradev replied:
changed to this ->
```
if c.hasColorMode && c.colorMode == mode {
		return
	}
```

hajimehoshi replied:
Why?

hajimehoshi replied:
This change is not necessary anyway. Revert this.

hajimehoshi commented on context.go:
```diff
@@ -99,10 +107,9 @@ func (c *Context) ColorMode() ColorMode {
+	if c.hasColorMode && c.colorMode == mode {
 		return
 	}
-
```
?

hajimehoshi replied:
This change is not necessary.

hajimehoshi commented on internal/theme/detect_darwin.go:
```diff
@@ -0,0 +1,22 @@
+//go:build darwin
+// +build darwin
```
We don't need this comments as the file suffix is already enough.

hajimehoshi replied:
Add a copyright comment for new files

```go
// SPDX-License-Identifier: Apache-2.0
// SPDX-FileCopyrightText: 2025 The Guigui Authors
```

hajimehoshi commented on internal/theme/detect_linux.go:
```diff
@@ -0,0 +1,20 @@
+//go:build linux
+// +build linux
```
ditto

hajimehoshi commented on internal/theme/detect_windows.go:
```diff
@@ -0,0 +1,28 @@
+//go:build windows
+// +build windows
```
ditto

hajimehoshi commented on internal/theme/theme.go:
```diff
@@ -0,0 +1,13 @@
+)
+
+func DetectSystemTheme() ColorMode {
+	return detectSystemTheme() // implemented in per-OS files
```
This would fail with other platforms like browsers. Add `theme_others.go` with build tags `//go:build !darwin && !linux && !windows`.

hajimehoshi replied:
Rename this to `systemColorMode`

hajimehoshi commented on internal/theme/detect_darwin.go:
```diff
@@ -0,0 +1,22 @@
+func detectSystemTheme() ColorMode {
+	out, err := exec.Command("defaults", "read", "-g", "AppleInterfaceStyle").Output()
+	if err != nil {
+		return ThemeLight
```
Isn't this ThemeUnknown?

hajimehoshi replied:
NVM

hajimehoshi commented on internal/theme/theme.go:
```diff
@@ -0,0 +1,13 @@
+package theme
```
Change the package name from theme to colormode, as theme sounds too general. Also rename the directory name.

hajimehoshi commented on internal/theme/theme.go:
```diff
@@ -0,0 +1,13 @@
+const (
+	ThemeLight ColorMode = iota
+	ThemeDark
+	ThemeUnknown
```
```go
const (
    Unknown ColorMode = iota
    Light
    Dark
)
```

hajimehoshi commented on internal/theme/theme.go:
```diff
@@ -0,0 +1,13 @@
+	ThemeUnknown
+)
+
+func DetectSystemTheme() ColorMode {
```
```go
func SystemColorMode() ColorMode {
```

hajimehoshi reviewed (commented):
Thanks!

hajimehoshi commented:
* What did you refer when you implement this?
* Have you check this worked on every platform?

hajimehoshi commented on context.go:
```diff
@@ -17,17 +18,24 @@ import (
 	case "dark":
 		defaultColorMode = ColorModeDark
 	default:
-		slog.Warn(fmt.Sprintf("invalid GUIGUI_COLOR_MODE: %s", mode))
```
Keep this panic as the `default` clause.

hajimehoshi replied:
This is not resolved yet.

hajimehoshi replied:
```go
case "":
    switch colormode.SystemColorMode() {
    case colormode.Light:
        defaultColorMode = ColorModeLight
    case colormode.Dark:
        defaultColorMode = ColorModeDark
    default:
        defaultColorMode = ColorModeLight
    }
default:
    slog.Warn(fmt.Sprintf("invalid GUIGUI_COLOR_MODE: %s", mode))
```

hajimehoshi commented on internal/theme/detect_darwin.go:
```diff
@@ -0,0 +1,22 @@
+	if strings.Contains(strings.ToLower(string(out)), "dark") {
+		return ThemeDark
+	}
+	fmt.Println(string(out))
```
?

temidaradev replied:
okay i deleted the println statement. Sorry for forgetting it in there 😄

hajimehoshi commented on internal/theme/detect_darwin.go:
```diff
@@ -0,0 +1,22 @@
+	}
+	fmt.Println(string(out))
+	return ThemeLight
+}
```
Optional: there is a more straightforward way to get the theme: https://github.com/libsdl-org/SDL/blob/795d1ae1fb44f1321c4bb323a237d01c69a55579/src/video/cocoa/SDL_cocoavideo.m#L247-L258 but using `default` command is also fine.

temidaradev replied:
Let's do this the way you want. 😄

temidaradev commented on context.go:
```diff
@@ -17,17 +18,24 @@ import (
 type ColorMode int
 
-var defaultColorMode ColorMode = ColorModeLight
+var defaultColorMode = ColorMode(theme.DetectSystemTheme())
```
``var defaultColorMode = ColorMode(colormode.SystemColorMode())``
this is for casting not calling it twice i guess...

hajimehoshi replied:
I mean, there are two calls for `colormode.SystemColorMode()` and why.

hajimehoshi replied:
As `defaultColorMode` is initialized in `init`, I don't think we need the call here.

```go
var defaultColorMode ColorMode
```

temidaradev commented:
> * What did you refer when you implement this?
> * Have you check this worked on every platform?
> 
> > I added checker for all three specific operating systems, checkers work properly because of special builds for each OS
> 
> * What is the 'checker'?

i used the word checker as "detector" but its just detecting system theme thats all 😃 I tested in both mac and windows so far but i will test again after applying all you requests, but i couldnt test on linux because of "X11/lib.h" error in NixOS.

temidaradev commented:
deleted the detect_linux.go for now

hajimehoshi commented on internal/colormode/theme_others.go:
```diff
@@ -0,0 +1,11 @@
+//go:build !darwin && !linux && !windows
```
Now we don't have a Linux implementation, right? Fix the build tag.

hajimehoshi replied:
Add the copyright comment.

hajimehoshi replied:
Rename this file to `colormode_others.go`

hajimehoshi commented on internal/colormode/theme_others.go:
```diff
@@ -0,0 +1,11 @@
+package colormode
+
+func systemColorMode() ColorMode {
+	// Unknown default to Light mode or what should I put in this function?
```
Remove this comment.

hajimehoshi commented on internal/colormode/theme.go:
```diff
@@ -0,0 +1,16 @@
+)
+
+func SystemColorMode() ColorMode {
+	return systemColorMode() // implemented in per-OS files
```
Remove the comment.

hajimehoshi commented on internal/colormode/detect_windows.go:
```diff
@@ -0,0 +1,28 @@
+// SPDX-License-Identifier: Apache-2.0
```
Rename this file to `colormode_wndows.go`.

hajimehoshi commented on internal/colormode/detect_darwin.go:
```diff
@@ -0,0 +1,20 @@
+// SPDX-License-Identifier: Apache-2.0
```
Rename this file to `colormode_darwin.go`.

hajimehoshi commented on internal/colormode/theme.go:
```diff
@@ -0,0 +1,16 @@
+// SPDX-License-Identifier: Apache-2.0
```
Rename this file to `colormode.go`

temidaradev commented on context.go:
```diff
@@ -17,17 +18,24 @@ import (
 		defaultColorMode = ColorModeLight
 	case "dark":
 		defaultColorMode = ColorModeDark
 	default:
```
i think this also fixed

temidaradev commented:
Mac testing:
https://github.com/user-attachments/assets/7cee2b5f-8637-47c8-a929-d9d6e6a19767

realskyquest commented:
I recommend for linux to check the  `gtk-application-prefer-dark-theme`, [GTK-Docs](https://docs.gtk.org/gtk4/property.Settings.gtk-application-prefer-dark-theme.html)

It is available in gtk3.0 and gtk4.0, it returns a `true` or `false`, by default a `false`, using `grep` which is a standard command on linux(*works in any distro*), works on KDE as well

```go
func systemColorMode() ColorMode {
	out, err := exec.Command("grep", "gtk-application-prefer-dark-theme", "~/.config/gtk-4.0/settings.ini").Output()
	if err != nil {
		return Light
	}
	if strings.Contains(strings.ToLower(string(out)), "true") {
		return Dark
	}
       out, err = exec.Command("grep", "gtk-application-prefer-dark-theme", "~/.config/gtk-3.0/settings.ini").Output()
	if err != nil {
		return Light
	}
	if strings.Contains(strings.ToLower(string(out)), "true") {
		return Dark
	}
	return Light
}
```

temidaradev commented:
tried linux but not working, i will find a solution later...

temidaradev commented:
> I recommend for linux to check the  `gtk-application-prefer-dark-theme`, [GTK-Docs](https://docs.gtk.org/gtk4/property.Settings.gtk-application-prefer-dark-theme.html)
> 
> 
> 
> It is available in gtk3.0 and gtk4.0, it returns a `true` or `false`, by default a `false`, using `grep` which is a standard command on linux(*works in any distro*), works on KDE as well
> 
> 
> 
> edit: added gtk-3.0 as fallback for older distros, usually it should never be executed
> 
> 
> 
> ```go
> 
> func systemColorMode() ColorMode {
> 
> 	out, err := exec.Command("grep", "gtk-application-prefer-dark-theme", "~/.config/gtk-4.0/settings.ini").Output()
> 
> 	if err != nil {
> 
> 		return Light
> 
> 	}
> 
> 	if strings.Contains(strings.ToLower(string(out)), "true") {
> 
> 		return Dark
> 
> 	}
> 
>        out, err = exec.Command("grep", "gtk-application-prefer-dark-theme", "~/.config/gtk-3.0/settings.ini").Output()
> 
> 	if err != nil {
> 
> 		return Light
> 
> 	}
> 
> 	if strings.Contains(strings.ToLower(string(out)), "true") {
> 
> 		return Dark
> 
> 	}
> 
> 	return Light
> 
> }
> 
> ```

Didnt see this... I will check this tomorrow.. but if the PR is mergable, Hoshi can merge it. I will specifically make linux.. Thanks for the source btw.

hajimehoshi commented on internal/colormode/colormode_windows.go:
```diff
@@ -0,0 +1,28 @@
+	if err != nil {
+		return Unknown
+	}
+	defer k.Close()
```
```go
defer func() {
    _ = k.Close()
}
```

is fine

hajimehoshi commented on internal/colormode/colormode_windows.go:
```diff
@@ -0,0 +1,28 @@
+package colormode
+
+import (
+	"golang.org/x/sys/windows/registry"
```
This is a new dependency. Run `go mod tidy` and this will update `go.mod`.

temidaradev commented:
Im so sorry because the PR needed a lot of reviews, Im sorry if i wasted your time by writing tons of reviews and comments. 🙏

hajimehoshi commented:
You don't have to apologize. Please address on what I commented. Thanks,

hajimehoshi commented on context.go:
```diff
@@ -11,22 +11,32 @@ import (
+	"github.com/hajimehoshi/guigui/internal/colormode"
 	"github.com/hajimehoshi/guigui/internal/locale"
+
+	"golang.org/x/text/language"
```
Do not swap the order.

```go
	"golang.org/x/text/language"

	"github.com/hajimehoshi/guigui/internal/colormode"
	"github.com/hajimehoshi/guigui/internal/locale"
```

temidaradev commented:
Okay i fixed the import order in context.go, used desired strings method to get dark mode in colormode_darwin.go and reverted this:
```
if c.hasColorMode && c.colorMode == mode {
		return
	}
```
to this:
```
if c.hasColorMode && mode == c.colorMode {
		return
	}

```
i copy and pasted the statement from main branch...

hajimehoshi reviewed (approved):
LGTM, thanks!
